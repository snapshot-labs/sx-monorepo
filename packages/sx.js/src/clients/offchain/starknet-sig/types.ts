export const baseDomain = {
  name: 'sx-starknet',
  version: '0.1.0'
};

export const domainTypes = {
  StarkNetDomain: [
    { name: 'name', type: 'felt' },
    { name: 'version', type: 'felt' },
    { name: 'chainId', type: 'felt' }
  ]
};

export const aliasTypes = {
  StarkNetDomain: domainTypes.StarkNetDomain,
  SetAlias: [
    { name: 'from', type: 'felt' },
    { name: 'alias', type: 'felt' },
    { name: 'timestamp', type: 'felt' }
  ]
};
