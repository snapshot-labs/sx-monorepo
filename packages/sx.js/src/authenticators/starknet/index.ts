import createEthSigAuthenticator from './ethSig';
import createEthTxAuthenticator from './ethTx';
import createStarkSigAuthenticator from './starkSig';
import createStarkTxAuthenticator from './starkTx';
import createVanillaAuthenticator from './vanilla';
import { Authenticator } from '../../clients/starknet/types';
import { NetworkConfig } from '../../types';
import { hexPadLeft } from '../../utils/encoding';

export function getAuthenticator(
  address: string,
  networkConfig: NetworkConfig
): Authenticator | null {
  const authenticator = networkConfig.authenticators[hexPadLeft(address)];
  if (!authenticator) return null;

  if (authenticator.type === 'vanilla') {
    return createVanillaAuthenticator();
  }

  if (authenticator.type === 'ethSig') {
    return createEthSigAuthenticator();
  }

  if (authenticator.type === 'ethTx') {
    return createEthTxAuthenticator();
  }

  if (authenticator.type === 'starkSig') {
    return createStarkSigAuthenticator();
  }

  if (authenticator.type === 'starkTx') {
    return createStarkTxAuthenticator();
  }

  return null;
}
