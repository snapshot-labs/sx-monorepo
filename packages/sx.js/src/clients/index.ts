export { EthereumTx } from './starknet/ethereum-tx';
export { EthereumSig } from './starknet/ethereum-sig';
export { StarknetTx } from './starknet/starknet-tx';
export { StarknetSig } from './starknet/starknet-sig';
export { L1Executor } from './starknet/l1-executor';
export { HerodotusController } from './starknet/herodotus';
export { EthereumTx as EvmEthereumTx } from './evm/ethereum-tx';
export { EthereumSig as EvmEthereumSig } from './evm/ethereum-sig';
export { EthereumTx as GovernorBravoEthereumTx } from './governor-bravo/ethereum-tx';
export { EthereumSig as OffchainEthereumSig } from './offchain/ethereum-sig';
export { StarknetSig as OffchainStarknetSig } from './offchain/starknet-sig';
