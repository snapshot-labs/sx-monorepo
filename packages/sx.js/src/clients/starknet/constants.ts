export const ESTIMATE_FEE_OVERHEAD_PERCENT = 50;
