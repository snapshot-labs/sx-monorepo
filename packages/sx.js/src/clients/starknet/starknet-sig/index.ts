import randomBytes from 'randombytes';
import {
  Account,
  CallData,
  shortString,
  typedData,
  uint256,
  validateAndParseAddress
} from 'starknet';
import {
  aliasTypes,
  baseDomain,
  proposeTypes,
  updateProposalTypes,
  voteTypes
} from './types';
import {
  Alias,
  ClientConfig,
  ClientOpts,
  Envelope,
  Propose,
  SignatureData,
  StarknetEIP712AliasMessage,
  StarknetEIP712ProposeMessage,
  StarknetEIP712UpdateProposalMessage,
  StarknetEIP712VoteMessage,
  UpdateProposal,
  Vote
} from '../../../types';
import { getStrategiesWithParams } from '../../../utils/strategies';

export class StarknetSig {
  config: ClientConfig & { manaUrl: string };

  constructor(opts: ClientOpts & { manaUrl: string }) {
    this.config = opts;
  }

  generateSalt() {
    return `0x${randomBytes(4).toString('hex')}`;
  }

  public async send(envelope: Envelope<Propose | UpdateProposal | Vote>) {
    const body = {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method: 'send',
        params: { envelope },
        id: null
      })
    };

    const res = await fetch(
      `${this.config.manaUrl}/stark_rpc/${this.config.networkConfig.eip712ChainId}`,
      body
    );
    const json = await res.json();

    return json.result;
  }

  public async sign<
    T extends
      | StarknetEIP712ProposeMessage
      | StarknetEIP712UpdateProposalMessage
      | StarknetEIP712VoteMessage
      | StarknetEIP712AliasMessage
  >(
    signer: Account,
    verifyingContract: string,
    message: T,
    types: any,
    primaryType: string
  ): Promise<SignatureData> {
    const domain = {
      ...baseDomain,
      chainId: this.config.networkConfig.eip712ChainId,
      verifyingContract
    };

    const data: typeof typedData.TypedData = {
      types,
      primaryType,
      domain,
      message
    };
    const signature = await signer.signMessage(data);

    return {
      address: validateAndParseAddress(signer.address),
      signature: Array.isArray(signature)
        ? signature.map(v => `0x${BigInt(v).toString(16)}`)
        : [`0x${signature.r.toString(16)}`, `0x${signature.s.toString(16)}`],
      message,
      domain,
      types,
      primaryType
    };
  }

  public async propose({
    signer,
    data
  }: {
    signer: Account;
    data: Propose;
  }): Promise<Envelope<Propose>> {
    const address = signer.address;

    const userStrategies = await getStrategiesWithParams(
      'propose',
      data.strategies,
      address,
      data,
      this.config
    );

    const message = {
      space: data.space,
      author: address,
      executionStrategy: {
        address: data.executionStrategy.addr,
        params: data.executionStrategy.params
      },
      userProposalValidationParams: CallData.compile({
        userStrategies
      }),
      metadataUri: shortString
        .splitLongString(data.metadataUri)
        .map(str => shortString.encodeShortString(str)),
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(
      signer,
      data.authenticator,
      message,
      proposeTypes,
      'Propose'
    );

    return {
      signatureData,
      data
    };
  }

  public async updateProposal({
    signer,
    data
  }: {
    signer: Account;
    data: UpdateProposal;
  }): Promise<Envelope<UpdateProposal>> {
    const address = signer.address;

    const message = {
      space: data.space,
      author: address,
      proposalId: uint256.bnToUint256(data.proposal),
      executionStrategy: {
        address: data.executionStrategy.addr,
        params: data.executionStrategy.params
      },
      metadataUri: shortString
        .splitLongString(data.metadataUri)
        .map(str => shortString.encodeShortString(str)),
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(
      signer,
      data.authenticator,
      message,
      updateProposalTypes,
      'UpdateProposal'
    );

    return {
      signatureData,
      data
    };
  }

  public async vote({
    signer,
    data
  }: {
    signer: Account;
    data: Vote;
  }): Promise<Envelope<Vote>> {
    const address = signer.address;

    const userVotingStrategies = await getStrategiesWithParams(
      'vote',
      data.strategies,
      address,
      data,
      this.config
    );

    const message = {
      space: data.space,
      voter: address,
      proposalId: uint256.bnToUint256(data.proposal),
      choice: `0x${data.choice.toString(16)}`,
      userVotingStrategies,
      metadataUri: shortString
        .splitLongString(data.metadataUri)
        .map(str => shortString.encodeShortString(str))
    };

    const signatureData = await this.sign(
      signer,
      data.authenticator,
      message,
      voteTypes,
      'Vote'
    );

    return {
      signatureData,
      data
    };
  }

  public async setAlias({
    signer,
    data
  }: {
    signer: Account;
    data: Alias;
  }): Promise<Envelope<Alias>> {
    const message = {
      from: validateAndParseAddress(signer.address),
      timestamp: parseInt((Date.now() / 1000).toFixed()),
      ...data
    };

    const signatureData = await this.sign(
      signer,
      '',
      message,
      aliasTypes,
      'SetAlias'
    );

    return {
      signatureData,
      data
    };
  }
}
