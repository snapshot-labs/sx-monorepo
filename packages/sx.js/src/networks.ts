export * from './evmNetworks';
export * from './starknetNetworks';
