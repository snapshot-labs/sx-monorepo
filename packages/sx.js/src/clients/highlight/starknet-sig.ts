import { Account, StarknetDomain, StarknetType, TypedData } from 'starknet';
import { Envelope, HighlightDomain, SetAlias } from './types';
import {
  HIGHLIGHT_STARKNET_DOMAIN,
  STARKNET_ALIASES_CONFIG
} from '../../highlightConstants';

export class HighlightStarknetSigClient {
  private highlightUrl: string;

  constructor(highlightUrl: string) {
    this.highlightUrl = highlightUrl;
  }

  public async getDomain(
    signer: Account,
    salt: bigint,
    to: string
  ): Promise<{
    starknetDomain: Required<StarknetDomain>;
    meta: Pick<HighlightDomain, 'salt' | 'verifyingContract'>;
  }> {
    return {
      starknetDomain: {
        ...HIGHLIGHT_STARKNET_DOMAIN,
        chainId: await signer.getChainId()
      },
      meta: {
        salt: `0x${salt.toString(16).padStart(64, '0')}`,
        verifyingContract: to
      }
    };
  }

  public async sign(
    signer: Account,
    domain: Required<StarknetDomain>,
    primaryType: string,
    types: Record<string, StarknetType[]>,
    message: Record<string, any>
  ): Promise<string> {
    const data: TypedData = {
      types,
      primaryType,
      domain,
      message
    };

    const signature = await signer.signMessage(data);

    return (
      Array.isArray(signature)
        ? signature.map(v => `0x${BigInt(v).toString(16)}`)
        : [`0x${signature.r.toString(16)}`, `0x${signature.s.toString(16)}`]
    ).join(',');
  }

  public async send(envelope: Envelope): Promise<any> {
    const { domain, message, primaryType, signer, signature } = envelope;

    const payload = {
      method: 'hl_postMessage',
      params: {
        domain,
        message,
        primaryType,
        signer,
        signature
      }
    };

    const body = {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(payload)
    };

    const res = await fetch(this.highlightUrl, body);

    if (!res.ok) {
      throw new Error(`Failed to send envelope: ${res.statusText}`);
    }

    const result = await res.json();

    if (result.error) {
      throw new Error(result.error);
    }

    return result;
  }

  public async setAlias({
    signer,
    data,
    salt
  }: {
    signer: Account;
    data: SetAlias;
    salt: bigint;
  }): Promise<Envelope> {
    const primaryType = 'SetAlias';

    const domain = await this.getDomain(
      signer,
      salt,
      STARKNET_ALIASES_CONFIG.address
    );

    const signature = await this.sign(
      signer,
      domain.starknetDomain,
      primaryType,
      STARKNET_ALIASES_CONFIG.types.setAlias,
      data
    );

    return {
      type: 'HIGHLIGHT_ENVELOPE',
      domain: { ...domain.starknetDomain, ...domain.meta },
      message: data,
      primaryType,
      signer: signer.address,
      signature
    };
  }
}
