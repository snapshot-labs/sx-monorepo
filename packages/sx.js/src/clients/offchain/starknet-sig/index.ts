import { Account, TypedData, validateAndParseAddress } from 'starknet';
import { aliasTypes, baseDomain } from './types';
import { OffchainNetworkStarknetConfig } from '../../../types';
import {
  EIP712SetAliasMessage,
  Envelope,
  SetAlias,
  SignatureData
} from '../types';

type StarknetSigClientOpts = {
  networkConfig: OffchainNetworkStarknetConfig;
};

export class StarknetSig {
  config: StarknetSigClientOpts;

  constructor(opts: StarknetSigClientOpts) {
    this.config = opts;
  }

  public async sign<T extends EIP712SetAliasMessage>(
    signer: Account,
    message: T,
    types: any,
    primaryType: string
  ): Promise<SignatureData> {
    const domain = {
      ...baseDomain,
      chainId: this.config.networkConfig.eip712ChainId
    };

    const data: TypedData = {
      types,
      primaryType,
      domain,
      message
    };
    const signature = await signer.signMessage(data);

    return {
      address: validateAndParseAddress(signer.address),
      signature: Array.isArray(signature)
        ? signature.map(v => `0x${BigInt(v).toString(16)}`)
        : [`0x${signature.r.toString(16)}`, `0x${signature.s.toString(16)}`],
      message,
      domain,
      types,
      primaryType
    };
  }

  public async setAlias({
    signer,
    data
  }: {
    signer: Account;
    data: SetAlias;
  }): Promise<Envelope<SetAlias>> {
    const message = {
      ...data,
      from: validateAndParseAddress(signer.address),
      timestamp: parseInt((Date.now() / 1000).toFixed())
    };

    const signatureData = await this.sign(
      signer,
      message,
      aliasTypes,
      'SetAlias'
    );

    return {
      signatureData,
      data
    };
  }
}
