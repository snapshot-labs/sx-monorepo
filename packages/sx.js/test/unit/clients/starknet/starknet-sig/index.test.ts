import { Account } from 'starknet';
import { afterAll, beforeAll, describe, expect, it, vi } from 'vitest';
import { StarknetSig } from '../../../../../src/clients/starknet/starknet-sig';
import { starknetNetworks, starknetSepolia } from '../../../../../src/networks';
import { starkProvider } from '../../../helpers';

describe('StarknetSig', () => {
  const address =
    '0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4';
  const privateKey = '0xcd613e30d8f16adf91b7584a2265b1f5';

  const account = new Account(starkProvider, address, privateKey);

  const client = new StarknetSig({
    starkProvider,
    networkConfig: starknetSepolia,
    ethUrl: 'https://rpc.brovider.xyz/5',
    manaUrl: 'https://mana.box',
    whitelistServerUrl: 'https://wls.snapshot.box'
  });

  beforeAll(() => {
    vi.spyOn(client, 'generateSalt').mockImplementation(() => '0x0');
  });

  afterAll(() => {
    vi.restoreAllMocks();
  });

  const { StarkSig } = starknetNetworks['sn-sep'].Authenticators;
  const { MerkleWhitelist } = starknetNetworks['sn-sep'].Strategies;

  it('should create propose envelope', async () => {
    const envelope = await client.propose({
      signer: account,
      data: {
        space:
          '0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8',
        authenticator: StarkSig,
        strategies: [
          {
            address: MerkleWhitelist,
            index: 0,
            params: '0x'
          }
        ],
        executionStrategy: {
          addr: '0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe',
          params: ['0x101']
        },
        metadataUri: 'ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca'
      }
    });

    expect(envelope).toMatchSnapshot();
  });

  it('should create update proposal envelope', async () => {
    const envelope = await client.updateProposal({
      signer: account,
      data: {
        space:
          '0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8',
        authenticator: StarkSig,
        executionStrategy: {
          addr: '0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe',
          params: ['0x101']
        },
        proposal: 1,
        metadataUri: 'ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca'
      }
    });

    expect(envelope).toMatchSnapshot();
  });

  it('should create vote envelope', async () => {
    const envelope = await client.vote({
      signer: account,
      data: {
        space:
          '0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8',
        authenticator: StarkSig,
        strategies: [
          {
            address: MerkleWhitelist,
            index: 0,
            params: '0x'
          }
        ],
        metadataUri: 'ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca',
        proposal: 1,
        choice: 1
      }
    });

    expect(envelope).toMatchSnapshot();
  });
});
