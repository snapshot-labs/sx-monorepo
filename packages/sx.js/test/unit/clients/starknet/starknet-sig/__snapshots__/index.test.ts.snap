// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StarknetSig > should create propose envelope 1`] = `
{
  "data": {
    "authenticator": "0x213bb25044b189ccfda9882999dba32e011781dc11b2a6efa2b3d232824378e",
    "executionStrategy": {
      "addr": "0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe",
      "params": [
        "0x101",
      ],
    },
    "metadataUri": "ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca",
    "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
    "strategies": [
      {
        "address": "0x13bcbe7318fb8aa219d264dcf5916feb873e596389ba93d923f9a23378cb743",
        "index": 0,
      },
    ],
  },
  "signatureData": {
    "address": "0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4",
    "message": {
      "author": "0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4",
      "executionStrategy": {
        "address": "0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe",
        "params": [
          "0x101",
        ],
      },
      "metadataUri": [
        "0x697066733a2f2f516d4e726d36784b75696231544874576b694e35434b7442",
        "0x4565725143447055746d6744716961553278446d6361",
      ],
      "salt": "0x0",
      "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
      "userProposalValidationParams": [
        "0",
      ],
    },
    "primaryType": "Propose",
    "signature": [
      "0x7faa90924da8c43edec0109655f719aa06d597a97286d2a4844bc072bffdf3f",
      "0x2cd44c86f6f223ca5080c6f6092b93a9c932825b1b6a65e0515debcb22742ea",
    ],
  },
}
`;

exports[`StarknetSig > should create update proposal envelope 1`] = `
{
  "data": {
    "authenticator": "0x213bb25044b189ccfda9882999dba32e011781dc11b2a6efa2b3d232824378e",
    "executionStrategy": {
      "addr": "0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe",
      "params": [
        "0x101",
      ],
    },
    "metadataUri": "ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca",
    "proposal": 1,
    "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
  },
  "signatureData": {
    "address": "0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4",
    "message": {
      "author": "0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4",
      "executionStrategy": {
        "address": "0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe",
        "params": [
          "0x101",
        ],
      },
      "metadataUri": [
        "0x697066733a2f2f516d4e726d36784b75696231544874576b694e35434b7442",
        "0x4565725143447055746d6744716961553278446d6361",
      ],
      "proposalId": {
        "high": "0x0",
        "low": "0x1",
      },
      "salt": "0x0",
      "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
    },
    "primaryType": "UpdateProposal",
    "signature": [
      "0x66f624dd2dcbc81ab4f6aea08db63b902a72517e461a4fbe30c27cd5b775e9b",
      "0x28993ed0dd85c5ac3b5017978f4251b014bc7bd375a02048c1e032997f75700",
    ],
  },
}
`;

exports[`StarknetSig > should create vote envelope 1`] = `
{
  "data": {
    "authenticator": "0x213bb25044b189ccfda9882999dba32e011781dc11b2a6efa2b3d232824378e",
    "choice": 1,
    "metadataUri": "ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca",
    "proposal": 1,
    "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
    "strategies": [
      {
        "address": "0x13bcbe7318fb8aa219d264dcf5916feb873e596389ba93d923f9a23378cb743",
        "index": 0,
      },
    ],
  },
  "signatureData": {
    "address": "0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4",
    "message": {
      "choice": "0x1",
      "metadataUri": [
        "0x697066733a2f2f516d4e726d36784b75696231544874576b694e35434b7442",
        "0x4565725143447055746d6744716961553278446d6361",
      ],
      "proposalId": {
        "high": "0x0",
        "low": "0x1",
      },
      "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
      "userVotingStrategies": [],
      "voter": "0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4",
    },
    "primaryType": "Vote",
    "signature": [
      "0xd64cf8eb4218e6b47ab1756ab850b82b5bb9eec760c8cd0c1b68bb2ea597be",
      "0x56a47d37a6980a7f095ac54b5ce5108325333c7642234fc677423a59a027e38",
    ],
  },
}
`;
